lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -453,15 +504,30 @@ impl<'s> SegmentHolder {
                 min_unsaved_version = min(min_unsaved_version, segment_persisted_version);
             }
 
-            max_persisted_version = max(max_persisted_version, segment_persisted_version)
+            max_persisted_version = max(max_persisted_version, segment_persisted_version);
+
+            // Release read-lock immediately after flush, explicit to make this more clear
+            drop(read_segment);
         }
+
         if has_unsaved {
             Ok(min_unsaved_version)
         } else {
             Ok(max_persisted_version)
         }
     }
 
+    /// Grab the RwLock's for all the given segment IDs.
+    fn segment_locks(
+        &self,
+        segment_ids: impl IntoIterator<Item = SegmentId>,
+    ) -> Vec<Arc<RwLock<dyn SegmentEntry>>> {
+        segment_ids
+            .into_iter()
+            .map(|segment_id| self.segments.get(&segment_id).unwrap().get())
```

Comment:
It is an internal function. The provided list of IDs is sourced internally from `self.segments` in [here](https://github.com/qdrant/qdrant/blob/82db55928d27e1d5ece77c38d27f6b57be8d2121/lib/collection/src/collection_manager/holders/segment_holder.rs#L440-L452).

So this can never panic in the current implementation.

Reply:
In this case I think it would be fitting to use`.expect()` with such a message. However, wouldn't we prefer to log and ignore?

Reply:
Implemented in 7c0d7060e7e7fa0e4d98a69064fe45bf925c605b

Reply:
Implemented in 7c0d7060e7e7fa0e4d98a69064fe45bf925c605b